In [3]:
import pandas as pd 
import numpy as np
from sklearn import preprocessing
data=pd.read_csv("final_data.csv")
#final_data is the final pre processed data

In [4]:
data.head()

Operator In Out Travelling Network Type  Rating Call Drop Category  \
0    Airtel            Indoor           4G       3       Satisfactory   
1  Vodafone            Indoor           4G       1       Call Dropped   
2    Airtel            Indoor           3G       1       Call Dropped   
3    Airtel            Indoor           4G       3       Satisfactory   
4    Airtel        Travelling           4G       3       Satisfactory   

   Latitude  Longitude   State Name  Average Data Speed Download  \
0  0.196228   0.325356    Karnataka                     0.271480   
1  0.632907   0.135432      Gujarat                     0.211519   
2  0.127093   0.290671   Tamil Nadu                     0.036321   
3  0.580181   0.155327      Gujarat                     0.175517   
4  0.425737   0.180503  Maharashtra                     0.333021   

   Average Data Speed Upload  Population  Literacy Rate  Rainfall  \
0                   0.197512           2              7         3   
1                   0.370879           2              8         2   
2                   0.062303           2              8         2   
3                   0.191200           2              8         2   
4                   0.395612           4              8         4   

   Area of the state  Average_Signal  
0           0.504087        0.476721  
1           0.518040        0.477252  
2           0.300607        0.478454  
3           0.518040        0.477699  
4           0.886194        0.476859

In [5]:
Vodafone=data[data['Operator']=='Vodafone']

In [6]:
Vodafone.count()

Operator                       5895
In Out Travelling              5895
Network Type                   5895
Rating                         5895
Call Drop Category             5895
Latitude                       5895
Longitude                      5895
State Name                     5895
Average Data Speed Download    5895
Average Data Speed Upload      5895
Population                     5895
Literacy Rate                  5895
Rainfall                       5895
Area of the state              5895
Average_Signal                 5895
dtype: int64

In [7]:
from sklearn import preprocessing
from numpy.linalg import inv
from sklearn.linear_model import LinearRegression

In [8]:
X=Vodafone[['Latitude','Longitude']]

In [9]:
X.count()

Latitude     5895
Longitude    5895
dtype: int64

In [10]:
y=Vodafone['Average Data Speed Download']

In [11]:
from sklearn.model_selection import train_test_split 
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
x_train.reset_index(inplace = True, drop = True) 
x_test.reset_index(inplace = True, drop = True) 
y_train.reset_index(inplace = True, drop = True) 
y_test.reset_index(inplace = True, drop = True) 
#We have used holdout method and 70%of data is taken as training data and rest as testing 

In [12]:
x_train.head()

Latitude  Longitude
0  0.587488   0.727512
1  0.769716   0.254796
2  0.769717   0.254796
3  0.449126   0.150402
4  0.842450   0.316010

In [13]:
n=x_train.count()[0]

In [14]:
n

4126

# Download Data Speed Latitude Longitude

In [15]:
sum_x1=0
sum_x2=0
sum_y1=0
sum_x1x1=0
sum_x1x2=0
sum_x2x2=0
sum_x1y1=0
sum_x2y1=0
for i in range(0,n):
    sum_x1=sum_x1+x_train['Latitude'][i]
    sum_x2=sum_x2+x_train['Longitude'][i]
    sum_y1=sum_y1+y_train[i]
    sum_x1x1=sum_x1x1+x_train['Latitude'][i]*x_train['Latitude'][i]
    sum_x1x2=sum_x1x2+x_train['Latitude'][i]*x_train['Longitude'][i]
    sum_x2x2=sum_x2x2+x_train['Longitude'][i]*x_train['Longitude'][i]
    sum_x1y1=sum_x1y1+x_train['Latitude'][i]*y_train[i]
    sum_x2y1=sum_x2y1+x_train['Longitude'][i]*y_train[i]
A = np.array([[n, sum_x1, sum_x2], [sum_x1, sum_x1x1, sum_x1x2],[sum_x2,sum_x1x2,sum_x2x2]])
B=np.array([sum_y1, sum_x1y1, sum_x2y1])

In [16]:
A

array([[4126.        , 2571.74145018, 1243.22590574],
       [2571.74145018, 1714.72844094,  788.61417653],
       [1243.22590574,  788.61417653,  550.93518414]])

In [17]:
B

array([540.00115372, 332.54638332, 164.91954805])

In [18]:
Inverse=inv(A)
w0=Inverse[0][0]*B[0]+Inverse[0][1]*B[1]+Inverse[0][2]*B[2]
w1=Inverse[1][0]*B[0]+Inverse[1][1]*B[1]+Inverse[1][2]*B[2]
w2=Inverse[2][0]*B[0]+Inverse[2][1]*B[1]+Inverse[2][2]*B[2]
W=np.array([w0,w1,w2])

In [19]:
W

array([ 0.14991164, -0.0380227 ,  0.01548413])

In [28]:
y_pred=w0+w1*x_test['Latitude']+w2*x_test['Longitude']
#y_pred gives the predicted values of y based on the regression model
MSE=0
for i in range(0,len(y_test)):
    MSE=MSE+(y_pred[i]-y_test[i])**2
RMSE=np.sqrt(MSE/len(y_test))
RMSE
#RMSE is used to find the goodness of the model and is calacuated as root of sum of square of difference of actual and prediacted values.

0.05225505598308186

In [29]:
ybar=y_test.mean()
ybar

0.12960855333182555

In [30]:
SSR=0
SST=0
SSE=0
y_test.count()

1769

In [31]:
for i in range (0,len(y_test)):
    SST=SST+(y_test[i]-ybar)*(y_test[i]-ybar)
    SSR=SSR+(y_pred[i]-ybar)*(y_pred[i]-ybar)
    SSE=SSE+(y_test[i]-y_pred[i])*(y_test[i]-y_pred[i])
    
    

In [32]:
print("Total sum of squares=", SST )

Total sum of squares= 4.897642655768034


In [33]:
print("Sum of squares due to regression =", SSR)

Sum of squares due to regression = 0.08510139175397098


In [34]:
print("Sum of square due to error=", SSE)

Sum of square due to error= 4.830415259281389


In [35]:
p=2
n=len(y_test)


Significance Test

F Test


In [36]:
W

array([ 0.14991164, -0.0380227 ,  0.01548413])

In [37]:
w0=W[0]
w0

0.14991164438273358

In [38]:
w1=W[1]
w2=W[2]


Regression line is y= bo + b1 * x1 + b2 * x2

b1 is coefficient of x1

b2 is coefficient of x2

H0 : b1=b2=0

Ha : One or both parameters not equal to zero

In [39]:
MSR=SSR/2
MSR

0.04255069587698549

In [40]:
MSE=SSE/(len(y_test)-3)
MSE

0.002735229478641783

In [41]:
F=MSR/MSE

In [42]:
F=MSR/MSE
#The F test is used to determine whether a significant relationship exists between the dependent variable and the set of all the independent variables; hence F test are referred to as  the test for overall significance

15.556536008859718

In [43]:
len(y_test)-3

1766

F alpha with numerator dof=2 and denominator dof=n-3(=1766) at level of significance=0.05 is equal to 3.0008

In [44]:
Falpha=3.00081980

In [45]:
F>=Falpha

True

So we reject H0. Now for b1 and b2 t tests are performed separately.


T test for b1

H0 : b1=0

Ha : b1 is not equal to zero

In [46]:
sum=0
m=x_test['Latitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Latitude'][i]-m)**2)
sum

50.85491563552573

In [47]:
import math
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb1=s/x

In [48]:
t=w1/sb1

In [49]:
t
#t distribution with n − p − 1 degrees of freedom.
#The t test is used to determine whether each of the individual independent variables is significant. A separate t test is conducted for each of the independent variables in the model; hence t tests are called as the test for individual significance. 

-5.184567700619688

In [50]:
len(y_test)-3

1766

So talpha=2.243316 at dof=len(y_test)-3 and alpha=0.025

In [51]:
talpha=2.243316
t<=-talpha

True

Hence we reject null hypothesis

T Test for b2

H0 : b2=0

Ha : b2 not equal to zero

In [52]:
sum=0
m=x_test['Longitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Longitude'][i]-m)**2)
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb2=s/x

In [53]:
t=w2/sb2

In [54]:
t

2.4487016715046606

In [55]:
t>=talpha

True

Hence we reject null hypothesis

# Hence we can conclude that Download data speed does depend on Latitude and Longitude significantly for Vodafone

# Upload Data Speed Latitude Longitude

Now doing the same for Data Speed Upload

In [56]:
n=x_train.count()[0]
y=Vodafone['Average Data Speed Upload']
X=Vodafone[['Latitude','Longitude']]

In [57]:
from sklearn.model_selection import train_test_split 
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
x_train.reset_index(inplace = True, drop = True) 
x_test.reset_index(inplace = True, drop = True) 
y_train.reset_index(inplace = True, drop = True) 
y_test.reset_index(inplace = True, drop = True) 

In [58]:
sum_x1=0
sum_x2=0
sum_y1=0
sum_x1x1=0
sum_x1x2=0
sum_x2x2=0
sum_x1y1=0
sum_x2y1=0
for i in range(0,n):
    sum_x1=sum_x1+x_train['Latitude'][i]
    sum_x2=sum_x2+x_train['Longitude'][i]
    sum_y1=sum_y1+y_train[i]
    sum_x1x1=sum_x1x1+x_train['Latitude'][i]*x_train['Latitude'][i]
    sum_x1x2=sum_x1x2+x_train['Latitude'][i]*x_train['Longitude'][i]
    sum_x2x2=sum_x2x2+x_train['Longitude'][i]*x_train['Longitude'][i]
    sum_x1y1=sum_x1y1+x_train['Latitude'][i]*y_train[i]
    sum_x2y1=sum_x2y1+x_train['Longitude'][i]*y_train[i]
A = np.array([[n, sum_x1, sum_x2], [sum_x1, sum_x1x1, sum_x1x2],[sum_x2,sum_x1x2,sum_x2x2]])
B=np.array([sum_y1, sum_x1y1, sum_x2y1])
Inverse=inv(A)
w0=Inverse[0][0]*B[0]+Inverse[0][1]*B[1]+Inverse[0][2]*B[2]
w1=Inverse[1][0]*B[0]+Inverse[1][1]*B[1]+Inverse[1][2]*B[2]
w2=Inverse[2][0]*B[0]+Inverse[2][1]*B[1]+Inverse[2][2]*B[2]
W=np.array([w0,w1,w2])

In [59]:
W


array([ 0.18425854, -0.01807794,  0.17552101])

In [60]:
y_pred=w0+w1*x_test['Latitude']+w2*x_test['Longitude']
#y_pred gives the predicted values of y based on the regression model
MSE=0
for i in range(0,len(y_test)):
    MSE=MSE+(y_pred[i]-y_test[i])**2
RMSE=np.sqrt(MSE/len(y_test))
RMSE
#RMSE is used to find the goodness of the model and is calacuated as root of sum of square of difference of actual and prediacted values.

0.12425533480624819

In [61]:
ybar=y_test.mean()
SSR=0
SST=0
SSE=0
for i in range (0,len(y_test)):
    SST=SST+(y_test[i]-ybar)*(y_test[i]-ybar)
    SSR=SSR+(y_pred[i]-ybar)*(y_pred[i]-ybar)
    SSE=SSE+(y_test[i]-y_pred[i])*(y_test[i]-y_pred[i])
print("Total sum of squares=", SST )
print("Sum of squares due to regression =", SSR)
print("Sum of square due to error=", SSE)


Total sum of squares= 29.095295708371616
Sum of squares due to regression = 2.0998000255972635
Sum of square due to error= 27.3122777750009


In [62]:
w0

0.18425853628088967

In [63]:
w1

-0.018077944351990477

In [64]:
w2

0.17552101131593134

So linear regression line is y=b0 + b1 * x1 + b2 * x2

b1 is coefficient of x1


b2 is coefficient of x2

H0 : b1=b2=0


Ha : One or both parameters not equal to zero

In [65]:
MSR=SSR/2
MSE=SSE/(len(y_test)-3)
F=MSR/MSE
Falpha=3.00081980
F

67.8860781175664

In [66]:
F>=Falpha

True

Hence we reject H0. As both parameters are not equal to zero so we will find out individually about each parameter through t test.

T test for b1

H0 : b1=0

Ha : b1 is not equal to zero

In [67]:
sum=0
m=x_test['Latitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Latitude'][i]-m)**2)
sum

50.85491563552573

In [68]:
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb1=s/x

In [69]:
sb1

0.017438821872945192

In [70]:
t= w1/sb1

In [71]:
t

-1.0366494069210506

talpha at dof = n-3 at 0.05/2=0.25 level of significance is 2.24

In [72]:
talpha=2.243316

In [73]:
t<=-talpha

False

Hence we accept the null hypothesis.

T test for b2

H0 : b2=0

Ha : b2 is not equal to zero

In [74]:
sum=0
m=x_test['Longitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Longitude'][i]-m)**2)
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb2=s/x

In [75]:
sb2

0.01503618006530918

In [76]:
t= w2/s

In [77]:
t

1.4113850140230828

In [78]:
t>=talpha

False

Hence we accept the null hypothesis.

Hence we can conclude that Data Speed Upload does not depend on Latitude and Longitude significantly for Vodafone.

# Download Signal Strength Latitude Longitude

In [80]:
n=x_train.count()[0]
y=Vodafone['Average_Signal']
X=Vodafone[['Latitude','Longitude']]
from sklearn.model_selection import train_test_split 
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
x_train.reset_index(inplace = True, drop = True) 
x_test.reset_index(inplace = True, drop = True) 
y_train.reset_index(inplace = True, drop = True) 
y_test.reset_index(inplace = True, drop = True) 
sum_x1=0
sum_x2=0
sum_y1=0
sum_x1x1=0
sum_x1x2=0
sum_x2x2=0
sum_x1y1=0
sum_x2y1=0
for i in range(0,n):
    sum_x1=sum_x1+x_train['Latitude'][i]
    sum_x2=sum_x2+x_train['Longitude'][i]
    sum_y1=sum_y1+y_train[i]
    sum_x1x1=sum_x1x1+x_train['Latitude'][i]*x_train['Latitude'][i]
    sum_x1x2=sum_x1x2+x_train['Latitude'][i]*x_train['Longitude'][i]
    sum_x2x2=sum_x2x2+x_train['Longitude'][i]*x_train['Longitude'][i]
    sum_x1y1=sum_x1y1+x_train['Latitude'][i]*y_train[i]
    sum_x2y1=sum_x2y1+x_train['Longitude'][i]*y_train[i]
A = np.array([[n, sum_x1, sum_x2], [sum_x1, sum_x1x1, sum_x1x2],[sum_x2,sum_x1x2,sum_x2x2]])
B=np.array([sum_y1, sum_x1y1, sum_x2y1])
Inverse=inv(A)
w0=Inverse[0][0]*B[0]+Inverse[0][1]*B[1]+Inverse[0][2]*B[2]
w1=Inverse[1][0]*B[0]+Inverse[1][1]*B[1]+Inverse[1][2]*B[2]
w2=Inverse[2][0]*B[0]+Inverse[2][1]*B[1]+Inverse[2][2]*B[2]
W=np.array([w0,w1,w2])
W

array([ 0.47789789,  0.00066354, -0.00199361])

In [90]:
y_pred=w0+w1*x_test['Latitude']+w2*x_test['Longitude']
#y_pred gives the predicted values of y based on the regression model
MSE=0
for i in range(0,len(y_test)):
    MSE=MSE+(y_pred[i]-y_test[i])**2
RMSE=np.sqrt(MSE/len(y_test))
RMSE
#RMSE is used to find the goodness of the model and is calacuated as root of sum of square of difference of actual and prediacted values.

0.0010971176828594544

In [91]:
ybar=y_test.mean()
SSR=0
SST=0
SSE=0
for i in range (0,len(y_test)):
    SST=SST+(y_test[i]-ybar)*(y_test[i]-ybar)
    SSR=SSR+(y_pred[i]-ybar)*(y_pred[i]-ybar)
    SSE=SSE+(y_test[i]-y_pred[i])*(y_test[i]-y_pred[i])
print("Total sum of squares=", SST )
print("Sum of squares due to regression =", SSR)
print("Sum of square due to error=", SSE)

Total sum of squares= 0.0023771228059470757
Sum of squares due to regression = 0.0002774763732627702
Sum of square due to error= 0.002129287294565887


So linear regression line is y=b0 + b1 * x1 + b2 * x2

F test

H0 : b1=b2=0


Ha : One or both parameters not equal to zero

In [92]:
MSR=SSR/2
MSE=SSE/(len(y_test)-3)
F=MSR/MSE
Falpha=3.00081980

In [93]:
F

115.06743980312828

In [94]:
F>=Falpha

True

Hence we reject H0. As both parameters are not equal to zero so we will find out individually about each parameter through t test.

T test for b1

H0 : b1=0


Ha : b1 is not equal to zero


In [95]:
sum=0
m=x_test['Latitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Latitude'][i]-m)**2)
sum

50.85491563552573

In [96]:
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb1=s/x

In [97]:
sb1

0.00015397680811755636

In [98]:
t=w1/sb1

In [99]:
t

4.309324576098681

In [100]:
w1

0.0006635360433702164

In [101]:
t>=talpha

True

Hence we reject null hypothesis

T test for b2

H0 : b2=0

Ha : b2 is not equal to zero

In [102]:
sum=0
m=x_test['Longitude'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Longitude'][i]-m)**2)
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb2=s/x
t=w2/sb2

In [103]:
t

-15.016321888875604

In [104]:
t<=-talpha

True

Hence we reject null hypothesis

# Hence we conclude that Average Signal Strength depend on Latitude and Longitude significantly for Vodafone